API_DESIGN.md

```diff
@@ -0,0 +1,217 @@
+But do not create a global `Context` message that is used across the whole API if there are different scenarios and different fields required for the context.  
+The same applies to messages that are returned by multiple resources.  
+For example, information about the `User` might be different when managing the user resource itself than when it's returned
+as part of an authorization or a manager role, where only limited information is needed.
+
+Prevent reusing messages for the creation and the retrieval of a resource.
+Returning messages might contain additional information that is not required or even not available for the creation of the resource.  
+What might sound obvious when designing the CreateUserRequest for example, where only an `organization_id` but not the 
+`organization_name` is available, might not be so obvious when designing some sub-resource like a user's `IdentityProviderLink`, 
+which might contain an `identity_provider_name` when returned but not when created.
+
+#### Operations and Methods
+
+Methods on a resource should be named using the following convention:
+- Create: `Create<resource>`
+- Update: `Update<resource>`
+- Delete: `Delete<resource>`
+- Get: `Get<resource>`
+- List: `List<resource>`
+- Search: `Search<resource>`
+
+Methods on a list of resources should be named using the following convention:
+- Add: `Add<resource>`
+- Remove: `Remove<resource>`
+- Set: `Set<resource>`
+
+## Error Handling
+
+The API returns machine-readable errors in the response body. This includes a status code, an error code and possibly some details.
+
+### Status Codes
+
+The API uses status codes to indicate the status of a request. Depending on the protocol used to call the API,
+the status code is returned as an HTTP status code or as a gRPC / connectRPC status code.
+Check the possible status codes https://zitadel.com/docs/apis/statuscodes
+
+### Error Codes
+
+Additionally to the status code, the API returns unique error codes for each type of error.
+The error codes are used to identify a specific error and can be used to handle the error programmatically.
+
+> [!NOTE]
+> Currently, ZITADEL might already return some error codes. However, they do not follow a specific pattern yet
+> and are not documented. We will update the error codes and document them in the future.
+
+### Error Message and Details
+
+The API returns additional details about the error in the response body.
+This includes a human-readable error message and additional information that can help the client to understand the error
+as well as machine-readable details that can be used to handle the error programmatically.
+Error details use the Google RPC error details format: https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto
+
+### Example
+
+HTTP/1.1 example:
+```
+HTTP/1.1 400 Bad Request
+Content-Type: application/json
+
+{
+  "code": "USER-001",
+  "message": "user requires a username",
```

Comment:
This message is very detailed, what would happen if there are multiple fields missing? IMO it would be ok to return a message like `input is invalid` because I can see the details below

Reply:
changed to a better example as discussed yesterday